src/t_zset.c

```diff
@@ -1452,8 +1452,29 @@ int zsetAdd(robj *zobj, double score, sds ele, int in_flags, int *out_flags, dou
 
             /* Remove and re-insert when score changed. */
             if (score != curscore) {
-                zobj->ptr = zzlDelete(zobj->ptr,eptr);
-                zobj->ptr = zzlInsert(zobj->ptr,ele,score);
+                unsigned char *sptr = lpNext(zobj->ptr, eptr);
+
+                /* If we are updating the score of the last element(i.e the one
+                 * with the biggest score) with bigger score, then it's 
+                 * enough to just replace its score. This will preserve the
+                 * ordering and save us some iterations over the listpack. */
+                int is_last = (lpNext(zobj->ptr, sptr) == NULL);
+                if (is_last && score > curscore) {
+                    char scorebuf[MAX_D2STRING_CHARS];
+                    int scorelen = 0;
+                    long long lscore;
+                    int score_is_long = double2ll(score, &lscore);
+                    if (!score_is_long)
+                        scorelen = d2string(scorebuf,sizeof(scorebuf),score);
+
+                    if (score_is_long)
+                        zobj->ptr = lpReplaceInteger(zobj->ptr, &sptr, lscore);
+                    else
+                        zobj->ptr = lpReplace(zobj->ptr, &sptr, (unsigned char*)scorebuf, scorelen);
+                } else {
+                    zobj->ptr = zzlDelete(zobj->ptr,eptr);
+                    zobj->ptr = zzlInsert(zobj->ptr,ele,score);
+                }
```

Comment:
How much benefit can this optimization bring? code become ugly.

Reply:
I'm thinking out loud and assuming random operations. Say we have 10 elements. Chance of zadd being called for the last element is ~%10 (if it is not an addition). Though, half of the time score will be less, then can we say it will speed up %5 of the (update) operations. If we have more elements, benefit will diminish. I'm not sure if it is not worth the extra code. Because it feels like we are optimizing for a very specific case only (hope the above reasoning is correct).

Reply:
I agree with both statements - the code indeed becomes ugly and it's questionable if the change is needed. Only reference I found is this - https://github.com/redis/redis/issues/5179 when the same optimization was added for the skiplist case of zset, but the code is way more cleaner there.

Will revert.